vcpkg - Add support for SFCGAL

ptitjano opened the pull request:
## Description

Add option to enable SFCGAL support in vcpkg build and enable it for windows and macos CI.

ptitjano committed:
vcpkg: Add feature to build with SFCGAL support

ptitjano committed:
ci(windows-qt6): Enable SFCGAL support

ptitjano committed:
ci(macos-qt6): Enable SFCGAL support

ptitjano committed:
qgssfcgalengine: Use source_location to replace __FUNCTION__ macros

`__FILE__`, `__FUNCTION__`, `__LINE__` are used by QgsSfcgalEngine to
display debug and error messages. However, it does not work with MSVC
compiler with the following error:

```
error C2457: '__FUNCTION__': predefined macro cannot appear outside of
a function body
```

This issue is fixed by using `source_location` from C++20.

m-kuhn commented on vcpkg/vcpkg.json:
```diff
@@ -2,7 +2,7 @@
     "default-registry": {
       "kind": "git",
-      "baseline": "9e7bfb1bcd1fa2e6e0755c4076441360b9c9ded7",
+      "baseline": "4ed84798137bcf664989fa432d41d278d7ad3b25",
```
Will conflict with https://github.com/qgis/QGIS/pull/63714

ptitjano committed:
cmake(SFCGAL): Use SFCGAL native config file if available

SFCGAL has been providing a cmake config module since version 2.1. By
calling `find_package(SFCGAL CONFIG)` at the beginning of
`FindSFCGAL.cmake`, this ensures to use this module if it is
available.
This means that the custom `FindSFCGAL.cmake` logic is only used with
SFCGAL 2.0.

ptitjano commented:
> Can you still give it a try (if this works we can have this as first approach plus a fallback)

The FindSFCGAL.cmake did not call `find_package(SFCGAL CONFIG)`. The cmake config module was never used. This should be fixed now.

ptitjano committed:
cmake(SFCGAL): Remove uneeded macos codepath

This will never be triggered because on macos/vcpkg, the SFCGAL
version is recent enough to have the system SFCGAL cmake module.

ptitjano commented on cmake/FindSFCGAL.cmake:
```diff
@@ -15,161 +15,165 @@
+        )
+      ENDIF (MSVC)
+
+    ELSEIF(APPLE AND QGIS_MAC_DEPS_DIR)
```
I have removed it.

ptitjano commented:
> @ptitjano this broke the build on Fedora 42 -- now I get:
> 
> ```
> In file included from /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalgeometry.h:31,
>                  from /home/nyall/dev/build-qgis/python/core/auto_generated/geometry/qgssfcgalgeometry.sip:24:
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:130:71: error: ‘source_location’ in namespace ‘std’ does not name a type
>   130 |       bool hasSucceedOrStack( QString *errorMsg = nullptr, const std::source_location &location = std::source_location::current() );
>       |                                                                       ^~~~~~~~~~~~~~~
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:130:66: note: ‘std::source_location’ is only available from C++20 onwards
>   130 |       bool hasSucceedOrStack( QString *errorMsg = nullptr, const std::source_location &location = std::source_location::current() );
>       |                                                                  ^~~
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:147:52: error: ‘source_location’ in namespace ‘std’ does not name a type
>   147 |       void addText( const QString &msg, const std::source_location &location = std::source_location::current() );
>       |                                                    ^~~~~~~~~~~~~~~
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:147:47: note: ‘std::source_location’ is only available from C++20 onwards
>   147 |       void addText( const QString &msg, const std::source_location &location = std::source_location::current() );
>       |                                               ^~~
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:130:104: error: ‘std::source_location’ has not been declared
>   130 |       bool hasSucceedOrStack( QString *errorMsg = nullptr, const std::source_location &location = std::source_location::current() );
>       |                                                                                                        ^~~~~~~~~~~~~~~
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:130:104: note: ‘std::source_location’ is only available from C++20 onwards
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:147:85: error: ‘std::source_location’ has not been declared
>   147 |       void addText( const QString &msg, const std::source_location &location = std::source_location::current() );
>       |                                                                                     ^~~~~~~~~~~~~~~
> /home/nyall/dev/QGIS/src/core/geometry/qgssfcgalengine.h:147:85: note: ‘std::source_location’ is only available from C++20 onwards
> ```
> 
> Can you repair please?

@nyalldawson Unfortunately, I don't understand where this comes from and I cannot reproduce it. If I understand the error message correctly, the compilation fails because `‘std::source_location’ ` is not found. Indeed, it is only available since C++20.

However, QGIS is built with C++20 by default  : [cmakelists](https://github.com/qgis/QGIS/blob/e52685c63c5b20345bafca08450a0235ce749172/CMakeLists.txt#L80)

I have tried to reproduce it locally with fedora 42 but it builds fine. I haven't done anything special except enabling SFCGAL support. Also both ubuntu:25.10 and fedora:42 CI pass and they have SFCGAL support enabled: [CI, fedora 42 logs](https://github.com/qgis/QGIS/actions/runs/19246699651/job/55022370386).

It seems to be working fine on my side and on the CI, maybe there's something different on your end?
